server/src/main/java/com/thoughtworks/go/server/dao/DataSharingSettingsSqlMapDao.java

```diff
@@ -48,11 +55,32 @@ public void saveOrUpdate(DataSharingSettings dataSharingSettings) throws Duplica
             existing = dataSharingSettings;
         }
 
-        sessionFactory.getCurrentSession().saveOrUpdate(existing);
+        DataSharingSettings toSave = existing;
+        transactionTemplate.execute(new TransactionCallbackWithoutResult() {
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus status) {
+                transactionSynchronizationManager.registerSynchronization(new TransactionSynchronizationAdapter() {
+                    @Override
+                    public void afterCommit() {
+                        invalidateCache();
+                    }
+                });
+
+                sessionFactory.getCurrentSession().saveOrUpdate(toSave);
+            }
+        });
     }
 
     public DataSharingSettings load() {
-        return transactionTemplate.execute(status -> (DataSharingSettings) sessionFactory.getCurrentSession().getNamedQuery("load.datasharing.settings").uniqueResult());
+        if (cachedSettings == null) {
```

Comment:
You need to synchronize on `cachedSettings` while setting as well as invalidating the cache, add double checking too.

Reply:
Ok. will make that change

Reply:
@ketan -- Data Sharing Settings API is used by Consent SPA, It might not be used so frequently, but its good to cache. 
Usage Data Reporting API Information is accessed by **every browser client every half an hour** to check if it can perform data reporting, hence usage data information is necessary to cache!